### PR TITLE
Test coverage

### PR DESCRIPTION
**Pull request by kryofly**
Please consider carefully the third and fourth patch before accepting.
I intend no logical change, only to split up large functions for better testability, and hopefully I did not introduce bugs.

### CODE REVIEW

**Comment by vertti** [comment]
@gcarq I think you should take a look at this, I would prefer to have the # FIX comments removed before merging but @kryofly needs answers before that.

File: freqtrade/analyze.py [hunk 1]
```diff
@@ -301,15 +301,20 @@ def get_signal(pair: str, signal: SignalType) -> bool:
...
 
     latest = dataframe.iloc[-1]
 
     # Check if dataframe is out of date
     signal_date = arrow.get(latest['date'])
     if signal_date < arrow.now() - timedelta(minutes=10):
+        logger.warning('Too old dataframe for pair %s', pair)
         return False
 
+    # FIX: 20180109, there could be some confusion because we will make a
```
Comment (gcarq) [inline freqtrade/analyze.py]: I'm not sure if I understand that correctly. `latest[signal.value] == 1` should yield in a bool. So it should be a already a bool in every case or not?

File: freqtrade/main.py [hunk 2]
```diff
@@ -120,6 +140,53 @@ def _process(nb_assets: Optional[int] = 0) -> bool:
     return state_changed
 
 
+# FIX: 20180110, why is cancel.order unconditionally here, whereas
```
Comment (gcarq) [inline freqtrade/main.py]: I'm not familiar with the timedout logic yet, I'll look into it

File: freqtrade/main.py [hunk 3]
```diff
@@ -133,35 +200,16 @@ def check_handle_timedout(timeoutvalue: int) -> None:
...
-                # if trade is not partially completed, just cancel the trade
-                exchange.cancel_order(trade.open_order_id)
-                trade.close_rate = None
-                trade.close_profit = None
-                trade.close_date = None
-                trade.is_open = True
-                trade.open_order_id = None
-                logger.info('Sell order timeout for %s.', trade)
+            if handle_timedout_limit_sell(trade, order):
+                # BUG? if there is more trades that are
```
Comment (gcarq) [inline freqtrade/main.py]: I will also look into that

File: freqtrade/main.py [hunk 4]
```diff
@@ -120,6 +140,53 @@ def _process(nb_assets: Optional[int] = 0) -> bool:
...
+        exchange.cancel_order(trade.open_order_id)
+        trade.close_rate = None
+        trade.close_profit = None
+        trade.close_date = None
+        trade.is_open = True
+        trade.open_order_id = None
+        logger.info('Sell order timeout for %s.', trade)
+        return True
+    else:
+        # TODO: figure out how to handle partially complete sell orders
```
Comment (gcarq) [inline freqtrade/main.py]: I would do it the same way it is done in `Trade.update()`, that means only considering orders which are closed. Then you don't need to handle partially orders

File: freqtrade/tests/test_analyze.py [hunk 5]
```diff
@@ -66,6 +68,43 @@ def test_returns_latest_sell_signal(mocker):
...
+def test_get_signal_execption_valueerror(mocker, caplog):
+    mocker.patch('freqtrade.analyze.get_ticker_history', return_value=1)
+    mocker.patch('freqtrade.analyze.analyze_ticker',
+                 side_effect=ValueError('xyz'))
+    assert not get_signal('foo', SignalType.BUY)
+    assert tt.log_has('Unable to analyze ticker for pair foo: xyz',
+                      caplog.record_tuples)
+
+
+# This error should never occur becase analyze_ticker is run first,
```
Comment (gcarq) [inline freqtrade/tests/test_analyze.py]: Good point, I also think this test is obsolete

File: freqtrade/tests/test_main.py [hunk 6]
```diff
@@ -384,6 +427,21 @@ def test_check_handle_timedout_buy(default_conf, ticker, limit_buy_order_old, mo
     assert len(trades) == 0
 
 
+def test_handle_timedout_limit_buy(default_conf, mocker):
+    cancel_order = MagicMock()
+    mocker.patch('freqtrade.exchange.cancel_order', cancel_order)
+    Trade.session = MagicMock()
+    trade = MagicMock()
+    order = {}
```
Comment (gcarq) [inline freqtrade/tests/test_main.py]: That can be simplified to:
```
order = {
    'remaining': 1,
    'amount': 1,
}
```

File: freqtrade/tests/test_main.py [hunk 7]
```diff
@@ -416,6 +474,21 @@ def test_check_handle_timedout_sell(default_conf, ticker, limit_sell_order_old,
     assert trade_sell.is_open is True
 
 
+def test_handle_timedout_limit_sell(default_conf, mocker):
+    cancel_order = MagicMock()
+    mocker.patch('freqtrade.exchange.cancel_order', cancel_order)
+    trade = MagicMock()
+    order = {}
```
Comment (gcarq) [inline freqtrade/tests/test_main.py]: same as above

File: freqtrade/main.py [hunk 8]
```diff
@@ -158,6 +158,53 @@ def _process(nb_assets: Optional[int] = 0) -> bool:
     return state_changed
 
 
+# FIX: 20180110, why is cancel.order unconditionally here, whereas
+#                it is conditionally called in the
+#                handle_timedout_limit_sell()?
+def handle_timedout_limit_buy(trade, order):
```
Comment (glonlas) [inline freqtrade/main.py]: Could you please define the type of parameters like in the rest of this file?
```python
def handle_timedout_limit_buy(trade: Trade, order: Dict) -> bool:
```

File: freqtrade/main.py [hunk 9]
```diff
@@ -158,6 +158,53 @@ def _process(nb_assets: Optional[int] = 0) -> bool:
...
+        # and close the order
+        trade.amount = order['amount'] - order['remaining']
+        trade.stake_amount = trade.amount * trade.open_rate
+        trade.open_order_id = None
+        logger.info('Partial buy order timeout for %s.', trade)
+        return False
+
+
+# FIX: 20180110, should cancel_order() be cond. or unconditionally called?
+def handle_timedout_limit_sell(trade, order):
```
Comment (glonlas) [inline freqtrade/main.py]:
```python
def handle_timedout_limit_sell(trade: Trade, order: Dict) -> bool:
```

File: freqtrade/main.py [hunk 10]
```diff
@@ -171,35 +218,16 @@ def check_handle_timedout(timeoutvalue: int) -> None:
...
-                logger.info('Sell order timeout for %s.', trade)
+            if handle_timedout_limit_sell(trade, order):
+                # BUG? if there is more trades that are
+                #      timed out, shouldn't we collect and
+                #      then return all of them?
+                #      Also the function signature is return None.
+                #      But we return True here.
                 return True
-            else:
-                # TODO: figure out how to handle partially complete sell orders
```
Comment (glonlas) [inline freqtrade/main.py]: This condition should be in your handle_timedout_limit_sell like for the buy.

File: freqtrade/main.py (hunk 4 above)
Comment (glonlas) [reply freqtrade/main.py]: I had multiple times experienced Sell trades that stay partially open. It means you have sold a part of crypto-currency and a remaining part is still in your bittrex wallet.

In this case, I recommend:
1. on the current Trade object to change the amount and stake amount with the value sold and close the trade
2. open a new trade with same buy value and date than the one we just closed except the amount/stake amount is the remaining value (value still in your bittrex wallet), and of course NULL for close_rate, close_profit, close_date, and open_order_id

Let me know if I am unclear.

File: freqtrade/tests/test_main.py [hunk 11]
```diff
@@ -50,6 +48,18 @@ def test_main_start_hyperopt(mocker):
     assert call_args.func is not None
 
 
+# def test_main_trader(mocker):
```
Comment (glonlas) [inline freqtrade/tests/test_main.py]: Could you please remove this commented test?

**Comment by glonlas** [comment]
This is a good PR, few comments and hint for the partial timeout sell logic

**Comment by vertti** [comment]
@glonlas @gcarq what's the situation with this, should we merge it?

**Pull request merged**